File: sudoku.js
let ans = [];

function vertical(board, row, col, k) {
    for (let i = 0; i < board.length; i++) {
        if (board[i][col] == k) return false;
    }
    return true;
}

function horizontal(board, row, col, k) {
    for (let i = 0; i < board[0].length; i++) {
        if (board[row][i] == k) return false;
    }
    return true;
}

function block(board, row, col, k) {
    let rblock = Math.floor(row / 3) * 3;
    let cblock = Math.floor(col / 3) * 3;
    for (let i = rblock; i < rblock + 3; i++) {
        for (let j = cblock; j < cblock + 3; j++) {
            if (board[i][j] == k) return false;
        }
    }
    return true;
}

function isValid(board, i, j, k) {
    return vertical(board, i, j, k) && horizontal(board, i, j, k) && block(board, i, j, k);
}

function sudokuHelper(board) {
    let n = board.length;
    let m = board[0].length;


    if (ans.length >= 10) return true;

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (board[i][j] == '.') {
                let flag=false;
                for (let k = '1'; k <= '9'; k++) {
                    if (isValid(board, i, j,k)) {
                        board[i][j] = k;
                        if (sudokuHelper(board)) {
                            flag=true;
                        }
                        board[i][j] = '.'; // Backtrack
                    }
                }
                return flag; // Return false if no valid number was found
            }
        }
    }
    let nBoard=board.map(ele=>Array.from(ele));
    ans.push(nBoard);
    return true; // Return true if the board is completely filled
}

function Sudoku(board){
    // let board = [
    //     ['5', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.'],
    //     ['.', '.', '.', '.', '.', '.', '.', '.', '.']
    // ];
    sudokuHelper(board);
    // ans.forEach(ele=>{
    //     ele.forEach(iele=>{
    //         console.log(iele.join(" "));
    //     })
    //     console.log("\n");
    // })
    return ans;
}

function isSudoku(board) { 
    let n = board.length;
    let m = board[0].length;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (board[i][j] != '.') {
                let temp = board[i][j];
                board[i][j] = '.'; // Temporarily empty the cell
                if (!isValid(board, i, j, temp)) {
                    board[i][j] = temp
                    return false;
                }
                board[i][j] = temp; // Restore the original value
            }
        }
    }
    return true;
}


module.exports={Sudoku,isSudoku};

File: C++ Logic/sudoku.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<vector<char>>>ans;

bool vertical(vector<vector<char>> &board, int row, int col, char k){
    int n = board.size();

    for (int i = 0; i < n; i++)
    {
        if (board[i][col] == k)
            return false;
    }
    return true;
}

bool horizontal(vector<vector<char>> &board, int row, int col, char k)
{
    int n = board[0].size();

    for (int i = 0; i < n; i++)
    {
        if (board[row][i] == k)
            return false;
    }
    return true;
}

bool block(vector<vector<char>> &board, int row, int col, char k)
{
    int rblock = row / 3 * 3;
    int cblock = col / 3 * 3;
    for (int i = rblock; i < rblock + 3; i++)
    {
        for (int j = cblock; j < cblock + 3; j++)
        {
            if (board[i][j] == k)
                return false;
        }
    }
    return true;
}

bool isValid(vector<vector<char>> &board, int i, int j, char k)
{
    return vertical(board, i, j, k) && horizontal(board, i, j, k) && block(board, i, j, k);
}

bool sudokuHelper(vector<vector<char>> board)
{
    int n = board.size();
    int m = board[0].size();
    if(ans.size()>=10)
        return true;

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            if (board[i][j] == '.')
            {
                bool flag=false;
                for (char k = '1'; k <= '9'; k++)
                {
                    if (isValid(board, i, j, k))
                    {
                        board[i][j] = k;
                        if (sudokuHelper(board)){
                            flag=true;
                        }
                    }
                    board[i][j] = '.';
                }
                return flag;
            }
        }
    }
    ans.push_back(board);
    return true;
}

int main()
{
    vector<vector<char>> board = {
        {'5', '6', '8', '9', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'},
        {'.', '.', '.', '.', '.', '.', '.', '.', '.'}
    };

    sudokuHelper(board);
    
    cout<<"size of ans is : "<<ans.size()<<endl<<endl;

    for(auto b:ans){
        for(int i=0;i<b.size();i++){
            for(int j=0;j<b[0].size();j++){
                cout<<b[i][j]<<" ";
            }
            cout<<"\n";
        }
        cout<<"\n\n";
    }

}